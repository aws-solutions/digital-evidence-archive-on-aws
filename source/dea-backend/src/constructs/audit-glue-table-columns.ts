/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import * as glue from '@aws-cdk/aws-glue-alpha';

const dynamoKeySchema = {
  name: 'key',
  type: glue.Schema.struct([
    {
      name: 'PK',
      type: glue.Schema.STRING,
    },
    {
      name: 'SK',
      type: glue.Schema.STRING,
    },
    {
      name: 'GSI1PK',
      type: glue.Schema.STRING,
    },
    {
      name: 'GSI1SK',
      type: glue.Schema.STRING,
    },
    {
      name: 'GSI2PK',
      type: glue.Schema.STRING,
    },
    {
      name: 'GSI2SK',
      type: glue.Schema.STRING,
    },
  ]),
};

export const auditGlueTableColumns = [
  {
    name: 'timestamp',
    type: glue.Schema.TIMESTAMP,
  },
  {
    name: 'logEventType',
    type: glue.Schema.STRING,
  },
  {
    name: 'userIdentity',
    type: glue.Schema.struct([
      {
        name: 'type',
        type: glue.Schema.STRING,
      },
      {
        name: 'userName',
        type: glue.Schema.STRING,
      },
      {
        name: 'sessionContext',
        type: glue.Schema.struct([
          {
            name: 'sessionIssuer',
            type: glue.Schema.struct([
              {
                name: 'userName',
                type: glue.Schema.STRING,
              },
            ]),
          },
        ]),
      },
    ]),
  },
  {
    name: 'actorIdentity',
    type: glue.Schema.struct([
      {
        name: 'idType',
        type: glue.Schema.STRING,
      },
      {
        name: 'sourceIp',
        type: glue.Schema.STRING,
      },
      {
        name: 'idPoolUserId',
        type: glue.Schema.STRING,
      },
      {
        name: 'username',
        type: glue.Schema.STRING,
      },
      {
        name: 'firstName',
        type: glue.Schema.STRING,
      },
      {
        name: 'lastName',
        type: glue.Schema.STRING,
      },
      {
        name: 'userUlid',
        type: glue.Schema.STRING,
      },
      {
        name: 'deaRole',
        type: glue.Schema.STRING,
      },
      {
        name: 'userPoolUserId',
        type: glue.Schema.STRING,
      },
      {
        name: 'authCode',
        type: glue.Schema.STRING,
      },
      {
        name: 'idToken',
        type: glue.Schema.STRING,
      },
    ]),
  },
  {
    // this is a datetime string, e.g. "2023-08-11T16:42:26Z"
    name: 'dateTime',
    type: glue.Schema.STRING,
  },
  {
    // this is a datetime string, e.g. "2023-08-11T16:42:26Z"
    name: 'eventTime',
    type: glue.Schema.STRING,
  },
  {
    name: 'eventType',
    type: glue.Schema.STRING,
  },
  {
    name: 'requestPath',
    type: glue.Schema.STRING,
  },
  {
    name: 'result',
    type: glue.Schema.STRING,
  },
  {
    name: 'eventName',
    type: glue.Schema.STRING,
  },
  {
    name: 'sourceComponent',
    type: glue.Schema.STRING,
  },
  {
    name: 'eventSource',
    type: glue.Schema.STRING,
  },
  {
    name: 'sourceIPAddress',
    type: glue.Schema.STRING,
  },
  {
    name: 'caseId',
    type: glue.Schema.STRING,
  },
  {
    name: 'fileId',
    type: glue.Schema.STRING,
  },
  {
    name: 'fileHash',
    type: glue.Schema.STRING,
  },
  {
    name: 'caseActions',
    type: glue.Schema.STRING,
  },
  {
    name: 'targetUserId',
    type: glue.Schema.STRING,
  },
  {
    name: 'eventID',
    type: glue.Schema.STRING,
  },
  {
    name: 'requestParameters',
    type: glue.Schema.struct([
      dynamoKeySchema,
      {
        name: 'requestItems',
        type: glue.Schema.array(glue.Schema.struct([dynamoKeySchema])),
      },
    ]),
  },
  {
    name: 'resources',
    type: glue.Schema.struct([
      {
        name: '0',
        type: glue.Schema.struct([
          {
            name: 'ARN',
            type: glue.Schema.STRING,
          },
        ]),
      },
    ]),
  },
];
