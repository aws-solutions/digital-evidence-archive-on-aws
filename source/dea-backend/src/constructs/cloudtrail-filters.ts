/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { FilterPattern } from 'aws-cdk-lib/aws-logs';

export const subscriptionFilter = FilterPattern.any(
  FilterPattern.exists('$.userIdentity.userName'),
  FilterPattern.exists('$.userIdentity.sessionContext.sessionIssuer.userName')
);
