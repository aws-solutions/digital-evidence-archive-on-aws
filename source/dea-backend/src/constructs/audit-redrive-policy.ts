/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Aws } from 'aws-cdk-lib';
import { CfnWorkGroup } from 'aws-cdk-lib/aws-athena';
import { ManagedPolicy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { CfnDeliveryStream } from 'aws-cdk-lib/aws-kinesisfirehose';
import { Key } from 'aws-cdk-lib/aws-kms';
import { LogGroup } from 'aws-cdk-lib/aws-logs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

export default function createAuditRedrivePolicy(
  scope: Construct,
  athenaWorkgroup: CfnWorkGroup,
  firehose: CfnDeliveryStream,
  auditLogGroup: LogGroup,
  trailLogGroup: LogGroup,
  queryResultBucket: Bucket,
  auditBucket: Bucket,
  glueDBName: string,
  glueTableName: string,
  kmsKey: Key
): ManagedPolicy {
  const redrivePolicy = new ManagedPolicy(scope, 'AuditRedrivePolicy', {
    description:
      'A Policy which can be attached to an IAM Role, allowing that role to run the Audit Redrive script.',
    statements: [
      new PolicyStatement({
        actions: [
          'cloudformation:ListExports',
          'athena:StartQueryExecution',
          'athena:GetQueryExecution',
          'athena:GetQueryResults',
        ],
        resources: ['*'],
      }),
      new PolicyStatement({
        actions: ['athena:GetWorkgroup'],
        resources: [
          `arn:${Aws.PARTITION}:athena:${Aws.REGION}:${Aws.ACCOUNT_ID}:workgroup/${athenaWorkgroup.name}`,
        ],
      }),
      new PolicyStatement({
        actions: ['firehose:PutRecordBatch'],
        resources: [
          `arn:${Aws.PARTITION}:firehose:${Aws.REGION}:${Aws.ACCOUNT_ID}:deliverystream/${firehose.ref}`,
        ],
      }),
      new PolicyStatement({
        actions: ['logs:DescribeLogStreams', 'logs:GetLogEvents'],
        resources: [auditLogGroup.logGroupArn, trailLogGroup.logGroupArn],
      }),
      new PolicyStatement({
        actions: [
          's3:GetBucketLocation',
          's3:GetObject',
          's3:ListBucket',
          's3:ListBucketMultipartUploads',
          's3:AbortMultipartUpload',
          's3:PutObject',
          's3:ListMultipartUploadParts',
        ],
        resources: [queryResultBucket.bucketArn, `${queryResultBucket.bucketArn}/*`],
      }),
      new PolicyStatement({
        actions: ['s3:GetObject'],
        resources: [`${auditBucket.bucketArn}/*`],
      }),
      new PolicyStatement({
        actions: ['s3:ListBucket'],
        resources: [auditBucket.bucketArn],
      }),
      new PolicyStatement({
        actions: ['glue:GetTable', 'glue:GetDatabase'],
        resources: [
          `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:catalog`,
          `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:database/${glueDBName}`,
          `arn:${Aws.PARTITION}:glue:${Aws.REGION}:${Aws.ACCOUNT_ID}:table/${glueDBName}/${glueTableName}`,
        ],
      }),
      new PolicyStatement({
        actions: ['kms:Encrypt', 'kms:Decrypt', 'kms:GenerateDataKey'],
        resources: [kmsKey.keyArn],
      }),
    ],
  });

  return redrivePolicy;
}
