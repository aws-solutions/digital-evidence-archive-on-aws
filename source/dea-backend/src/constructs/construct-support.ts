/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { CfnOutput, CfnOutputProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { deaConfig } from '../config';

// methods created to avoid linting/sonarqube errors related to unused objects (eslint no-new)
export const createCfnOutput = (scope: Construct, id: string, props: CfnOutputProps): CfnOutput => {
  return new CfnOutput(
    scope,
    id,
    Object.assign(
      {},
      {
        ...props,
        exportName: `${deaConfig.stage()}-${id}`
      }
    )
  );
};
