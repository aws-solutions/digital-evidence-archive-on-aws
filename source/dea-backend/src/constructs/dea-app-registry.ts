/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Application } from '@aws-cdk/aws-servicecatalogappregistry-alpha';
import { Aws, CfnMapping, Fn, Stack, Tags } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface DeaAppRegisterProps {
  solutionId: string;
  solutionName: string;
  solutionVersion: string;
  appRegistryApplicationName: string;
  applicationType: string;
}

export class DeaAppRegisterConstruct extends Construct {
  readonly registryApplication: Application;
  private readonly appRegMap: CfnMapping;

  constructor(stack: Stack, id: string, props: DeaAppRegisterProps) {
    super(stack, id);
    this.appRegMap = this.createMapForAppRegistry(stack, props);
    this.registryApplication = this.createAppRegistry(stack);
    this.createAttributeGroup();
    this.applyTagsToApplication();
  }

  private createAppRegistry(stack: Stack): Application {
    const application = new Application(stack, 'AppRegistry', {
      applicationName: Fn.join('-', [
        this.appRegMap.findInMap('Data', 'AppRegistryApplicationName'),
        Aws.REGION,
        Aws.ACCOUNT_ID,
        Aws.STACK_NAME, // AWS CloudFormation has an overall limit of 128 characters for the name of a stack.
      ]),
      description: 'Service Catalog application to track and manage all your resources for the solution',
    });
    application.associateApplicationWithStack(stack);

    return application;
  }

  private createAttributeGroup() {
    this.registryApplication.addAttributeGroup('DefaultApplicationAttributes', {
      attributeGroupName: Fn.join('-', [Aws.REGION, Aws.STACK_NAME]),
      description: 'Attribute group for solution information',
      attributes: {
        applicationType: this.appRegMap.findInMap('Data', 'ApplicationType'),
        version: this.appRegMap.findInMap('Data', 'SolutionVersion'),
        solutionID: this.appRegMap.findInMap('Data', 'ID'),
        solutionName: this.appRegMap.findInMap('Data', 'SolutionName'),
      },
    });
  }

  private createMapForAppRegistry(stack: Stack, props: DeaAppRegisterProps) {
    const map = new CfnMapping(stack, 'Solution');
    map.setValue('Data', 'ID', props.solutionId);
    map.setValue('Data', 'SolutionVersion', props.solutionVersion);
    map.setValue('Data', 'AppRegistryApplicationName', props.appRegistryApplicationName);
    map.setValue('Data', 'SolutionName', props.solutionName);
    map.setValue('Data', 'ApplicationType', props.applicationType);

    return map;
  }

  private applyTagsToApplication() {
    Tags.of(this.registryApplication).add('Solutions:SolutionID', this.appRegMap.findInMap('Data', 'ID'));
    Tags.of(this.registryApplication).add(
      'Solutions:SolutionName',
      this.appRegMap.findInMap('Data', 'SolutionName')
    );
    Tags.of(this.registryApplication).add(
      'Solutions:SolutionVersion',
      this.appRegMap.findInMap('Data', 'SolutionVersion')
    );
    Tags.of(this.registryApplication).add(
      'Solutions:ApplicationType',
      this.appRegMap.findInMap('Data', 'ApplicationType')
    );
  }
}
