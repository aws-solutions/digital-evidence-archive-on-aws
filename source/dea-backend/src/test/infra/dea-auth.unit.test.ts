/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import * as cdk from 'aws-cdk-lib';
import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import 'source-map-support/register';
import { DeaAuthConstruct } from '../../constructs/dea-auth';

describe('DeaBackend constructs', () => {
  afterAll(() => {
    delete process.env.STAGE;
  });

  it('synthesizes the way we expect', () => {
    process.env.STAGE = 'test';
    const app = new cdk.App();
    const stack = new Stack(app, 'test-stack');

    const apiEndpointArns = new Map([
      ['A', 'Aarn'],
      ['B', 'Barn'],
      ['C', 'Carn'],
      ['D', 'Darn'],
    ]);
    new DeaAuthConstruct(stack, 'DeaAuth', { apiEndpointArns });

    // Prepare the stack for assertions.
    const template = Template.fromStack(stack);

    expect.addSnapshotSerializer({
      test: (val) => typeof val === 'string' && val.includes('zip'),
      print: (val) => {
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        const newVal = (val as string).replace(/([A-Fa-f0-9]{64})/, '[HASH REMOVED]');
        return `"${newVal}"`;
      },
    });

    expect(template).toMatchSnapshot();
  });

  it('works with a domain config', () => {
    process.env.STAGE = 'demo';

    const app = new cdk.App();
    const stack = new Stack(app, 'test-stack');

    const apiEndpointArns = new Map([
      ['A', 'Aarn'],
      ['B', 'Barn'],
      ['C', 'Carn'],
      ['D', 'Darn'],
    ]);
    new DeaAuthConstruct(stack, 'DeaAuth', { apiEndpointArns });

    // Prepare the stack for assertions.
    const template = Template.fromStack(stack);

    template.hasResourceProperties('AWS::Cognito::IdentityPool', {
      AllowUnauthenticatedIdentities: false,
    });
  });
});
