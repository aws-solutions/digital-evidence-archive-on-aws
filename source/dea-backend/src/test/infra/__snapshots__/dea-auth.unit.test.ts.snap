// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeaBackend constructs synthesizes the way we expect 1`] = `
Object {
  "Outputs": Object {
    "DeaAuthidentityPoolIdEBA0D0C1": Object {
      "Export": Object {
        "Name": "[STAGE-REMOVED]-identityPoolId",
      },
      "Value": Object {
        "Ref": "DeaAuthDEAIdentityPool0B891C1B",
      },
    },
    "DeaAuthpoolProviderUrl9C814957": Object {
      "Export": Object {
        "Name": "[STAGE-REMOVED]-poolProviderUrl",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "DeaAuthDEAUserPool22DB2BAB",
          "ProviderURL",
        ],
      },
    },
    "DeaAuthuserPoolClientId8B436482": Object {
      "Export": Object {
        "Name": "[STAGE-REMOVED]-userPoolClientId",
      },
      "Value": Object {
        "Ref": "DeaAuthDEAUserPooldeaappclient3742F9D7",
      },
    },
    "DeaAuthuserPoolIdB0FAC1EB": Object {
      "Export": Object {
        "Name": "[STAGE-REMOVED]-userPoolId",
      },
      "Value": Object {
        "Ref": "DeaAuthDEAUserPool22DB2BAB",
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "DeaAuthAuthTestGroup72A5E3A7": Object {
      "Properties": Object {
        "Description": "Group used for auth e2e [STAGE-REMOVED]ing",
        "GroupName": "AuthTestGroup",
        "Precedence": 100,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "DeaAuthAuthTestGroupRole23D2E7FA",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "DeaAuthAuthTestGroupRole23D2E7FA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role used for auth e2e [STAGE-REMOVED]ing",
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthAuthTestGroupRolePolicy8F41C7CD": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeaAuthAuthTestGroupRolePolicy8F41C7CD",
        "Roles": Array [
          Object {
            "Ref": "DeaAuthAuthTestGroupRole23D2E7FA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeaAuthCaseWorkerGroupE79505D6": Object {
      "Properties": Object {
        "Description": "Group containing users who need access to case APIs",
        "GroupName": "CaseWorkerGroup",
        "Precedence": 1,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "DeaAuthCaseWorkerGroupRole388C20D0",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "DeaAuthCaseWorkerGroupRole388C20D0": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role containing users who need access to case APIs",
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthCaseWorkerGroupRolePolicy37058665": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeaAuthCaseWorkerGroupRolePolicy37058665",
        "Roles": Array [
          Object {
            "Ref": "DeaAuthCaseWorkerGroupRole388C20D0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeaAuthCreateCasesTestGroup7CEF506E": Object {
      "Properties": Object {
        "Description": "Group used for create cases API e2e [STAGE-REMOVED]ing",
        "GroupName": "CreateCasesTestGroup",
        "Precedence": 100,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "DeaAuthCreateCasesTestGroupRoleB1D8969D",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "DeaAuthCreateCasesTestGroupRoleB1D8969D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role used for create cases API e2e [STAGE-REMOVED]ing",
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthCreateCasesTestGroupRolePolicy933145EF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeaAuthCreateCasesTestGroupRolePolicy933145EF",
        "Roles": Array [
          Object {
            "Ref": "DeaAuthCreateCasesTestGroupRoleB1D8969D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeaAuthDEAIdentityPool0B891C1B": Object {
      "Properties": Object {
        "AllowClassicFlow": false,
        "AllowUnauthenticatedIdentities": false,
        "CognitoIdentityProviders": Array [
          Object {
            "ClientId": Object {
              "Ref": "DeaAuthDEAUserPooldeaappclient3742F9D7",
            },
            "ProviderName": Object {
              "Fn::GetAtt": Array [
                "DeaAuthDEAUserPool22DB2BAB",
                "ProviderName",
              ],
            },
            "ServerSideTokenCheck": true,
          },
        ],
      },
      "Type": "AWS::Cognito::IdentityPool",
    },
    "DeaAuthDEAUserPool22DB2BAB": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_email",
              "Priority": 1,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": true,
          "InviteMessageTemplate": Object {
            "EmailMessage": "Hello {username}, you have been invited to use DEA! Your temporary password is {####}",
            "EmailSubject": "Digital Evidence Archive Signup Invitation",
            "SMSMessage": "Hello {username}, your temporary password for our DEA is {####}",
          },
        },
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": Object {
          "PasswordPolicy": Object {
            "MinimumLength": 8,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true,
            "TemporaryPasswordValidityDays": 1,
          },
        },
        "Schema": Array [
          Object {
            "Mutable": false,
            "Name": "family_name",
            "Required": true,
          },
          Object {
            "Mutable": false,
            "Name": "given_name",
            "Required": true,
          },
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "DeaAuthDEAUserPooldeaappclient3742F9D7": Object {
      "Properties": Object {
        "AccessTokenValidity": 720,
        "AllowedOAuthFlows": Array [
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": Array [
          "https://example.com",
        ],
        "ClientName": "dea-app-client",
        "EnableTokenRevocation": true,
        "ExplicitAuthFlows": Array [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_ADMIN_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "GenerateSecret": false,
        "IdTokenValidity": 720,
        "PreventUserExistenceErrors": "ENABLED",
        "RefreshTokenValidity": 720,
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "TokenValidityUnits": Object {
          "AccessToken": "minutes",
          "IdToken": "minutes",
          "RefreshToken": "minutes",
        },
        "UserPoolId": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "DeaAuthGetCaseTestGroup01C4525C": Object {
      "Properties": Object {
        "Description": "Group used for get cases API e2e [STAGE-REMOVED]ing",
        "GroupName": "GetCaseTestGroup",
        "Precedence": 100,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "DeaAuthGetCaseTestGroupRoleF15B4661",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "DeaAuthGetCaseTestGroupRoleF15B4661": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role used for get cases API e2e [STAGE-REMOVED]ing",
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthGetCaseTestGroupRolePolicy852A4EA7": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "execute-api:Invoke",
              "Effect": "Allow",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DeaAuthGetCaseTestGroupRolePolicy852A4EA7",
        "Roles": Array [
          Object {
            "Ref": "DeaAuthGetCaseTestGroupRoleF15B4661",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DeaAuthIdPoolAuthenticatedRole3AD16775": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthIdPoolUnAuthenticatedRoleA427A1FB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "unauthenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "DeaAuthDEAIdentityPool0B891C1B",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "DeaAuthIdentityPoolCognitoRoleAttachment914C22D9": Object {
      "Properties": Object {
        "IdentityPoolId": Object {
          "Ref": "DeaAuthDEAIdentityPool0B891C1B",
        },
        "RoleMappings": Object {
          "roleMappingsKey": Object {
            "AmbiguousRoleResolution": "AuthenticatedRole",
            "IdentityProvider": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "cognito-idp.us-east-1.amazonaws.com/",
                  Object {
                    "Ref": "DeaAuthDEAUserPool22DB2BAB",
                  },
                  ":",
                  Object {
                    "Ref": "DeaAuthDEAUserPooldeaappclient3742F9D7",
                  },
                ],
              ],
            },
            "RulesConfiguration": Object {
              "Rules": Array [
                Object {
                  "Claim": "cognito:groups",
                  "MatchType": "Contains",
                  "RoleARN": Object {
                    "Fn::GetAtt": Array [
                      "DeaAuthCaseWorkerGroupRole388C20D0",
                      "Arn",
                    ],
                  },
                  "Value": "CaseWorkerGroup",
                },
                Object {
                  "Claim": "cognito:groups",
                  "MatchType": "Contains",
                  "RoleARN": Object {
                    "Fn::GetAtt": Array [
                      "DeaAuthAuthTestGroupRole23D2E7FA",
                      "Arn",
                    ],
                  },
                  "Value": "AuthTestGroup",
                },
                Object {
                  "Claim": "cognito:groups",
                  "MatchType": "Contains",
                  "RoleARN": Object {
                    "Fn::GetAtt": Array [
                      "DeaAuthCreateCasesTestGroupRoleB1D8969D",
                      "Arn",
                    ],
                  },
                  "Value": "CreateCasesTestGroup",
                },
                Object {
                  "Claim": "cognito:groups",
                  "MatchType": "Contains",
                  "RoleARN": Object {
                    "Fn::GetAtt": Array [
                      "DeaAuthGetCaseTestGroupRoleF15B4661",
                      "Arn",
                    ],
                  },
                  "Value": "GetCaseTestGroup",
                },
              ],
            },
            "Type": "Token",
          },
        },
        "Roles": Object {
          "authenticated": Object {
            "Fn::GetAtt": Array [
              "DeaAuthIdPoolAuthenticatedRole3AD16775",
              "Arn",
            ],
          },
          "unauthenticated": Object {
            "Fn::GetAtt": Array [
              "DeaAuthIdPoolUnAuthenticatedRoleA427A1FB",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "DeaAuthuserpoolclientidssmparam605B45CC": Object {
      "Properties": Object {
        "AllowedPattern": ".*",
        "Description": "stores the user pool client id for use in token verification on the backend",
        "Name": "/dea/us-east-1/[STAGE-REMOVED]-userpool-client-id-param",
        "Tier": "Standard",
        "Type": "String",
        "Value": Object {
          "Ref": "DeaAuthDEAUserPooldeaappclient3742F9D7",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "DeaAuthuserpoolidssmparamE2094B6A": Object {
      "Properties": Object {
        "AllowedPattern": ".*",
        "Description": "stores the user pool id for use in token verification on the backend",
        "Name": "/dea/us-east-1/[STAGE-REMOVED]-userpool-id-param",
        "Tier": "Standard",
        "Type": "String",
        "Value": Object {
          "Ref": "DeaAuthDEAUserPool22DB2BAB",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
